### DISCUSSION TITLE
Technical Discussion

### DISCUSSION THREAD

**Discussion by T145** [T145/black-mirror#76] [General]
Discuss all coding related issues here. This can be anything from asking general questions about how things work, giving feedback, improving list processing, developing lists, etc.

**Comment by T145** [comment]
@iam-py-test @DandelionSprout Would either of you be able to help me out w/ understanding how uBlock, Easylist, and Adguard interpret their filters? At first glance it seemed like one filter would be sufficient for all three, and that the `||example.com^` syntax means that `example.com` is fully blocked.

I'm thinking that a separate filter in my program would be needed for uBlock, since it seems `example.com*` means the same thing as the aforementioned syntax. I've listed some used references below.

- https://github.com/uBlockOrigin/uBlock-issues/wiki/Static-filter-syntax#extended-syntax
- https://kb.adguard.com/en/general/dns-filtering-syntax
- https://kb.adguard.com/en/general/how-to-create-your-own-ad-filters#examples-1

**Reply by iam-py-test** [reply]
I'm far from being an expert, but this is what seems to happen:
- example.com*:

uBlock Origin: Example.com and test.example.com is blocked both as document and third-party, but not testexample.com or example.org. Example.comm is only blocked as a subresource. 
AdGuard: Example.com, test.example.com, testexample.com,  example.comm load (well, they don't exist but aren't blocked) but all subresources are blocked.

- ||example.com^

uBlock Origin: example.com and test.example.com blocked, but nothing else
AdGuard: Sub resources from example.com and test.example.com blocked but nothing else

**Reply by DandelionSprout** [reply]
The linked-to guides are pretty poorly written and confusing, and https://github.com/DandelionSprout/adfilt/blob/master/Wiki/SyntaxMeaningsThatAreActuallyHumanReadable.md is of much better assistance.

`||example.com^` is supposed to block the sites on all Top 3 adblockers, with the added benefit in uBO of showing a warning before visiting the page.

In the very, *very* rare scenario that someone would want to block the site on all Top 3 adblockers *without* showing a warning in uBO, something like `^example.com^` would've been preferable to `example.com*` (Edit: Seems like uBO's warning prompts recognise the use of prefix `^` and `*` too nowadays.)

**Reply by T145** [reply]
Ok, thanks; so as far as I understand these are some ways to block `example.com` in each format:

```
||example.com^
||example.com^$third-party
||example.com^$~third-party
||example.com^$popup,third-party
||example.com/js/webscript.js$script,third-party # blocks "webscript.js" from example.com
||site.com/*/*.jpg$image,domain=example.com # blocks all JPEG images from example.com on site.com
||site.com/js/getlinkz.js$script,domain=example.com # blocks "getlinkz.js" from example.com on site.com
.gif$image,domain=example.com # blocks all GIF images from example.com
```

In my mind each of these listings warrant `example.com` being added to my blacklist.

UBO supports a hosts file format for blocking hosts, and uses some other unique identifiers like `$3p` or `$1p`. I don't think a whole separate filter would need to be included in my program to handle it though: some more regex cases would be all that's required.

Are there any I'm missing?

**Reply by iam-py-test** [reply]
Don't think so...
